## 라이브러리 & 데이터 준비

In [1]:
import numpy as np
import pandas as pd

In [2]:
# raw data file 총 2개
data = pd.read_csv('data.csv')
bus_bts = pd.read_csv('bus_bts.csv')

## 첫번째 data

### data 기본정보

In [3]:
data.shape

(415423, 21)

데이터 행 개수 : 415423  
데이터 열 개수 : 21개

In [4]:
data.head()

id        date  bus_route_id in_out  station_code  station_name  latitude  \
0   0  2019-09-01       4270000     시외           344         제주썬호텔  33.48990   
1   1  2019-09-01       4270000     시외           357          한라병원  33.48944   
2   2  2019-09-01       4270000     시외           432          정존마을  33.48181   
3   3  2019-09-01       4270000     시내          1579  제주국제공항(600번)  33.50577   
4   4  2019-09-01       4270000     시내          1646      중문관광단지입구  33.25579   

   longitude  6~7_ride  7~8_ride  ...  9~10_ride  10~11_ride  11~12_ride  \
0  126.49373       0.0       1.0  ...        5.0         2.0         6.0   
1  126.48508       1.0       4.0  ...        2.0         5.0         6.0   
2  126.47352       1.0       1.0  ...        2.0         0.0         0.0   
3  126.49252       0.0      17.0  ...       26.0        14.0        16.0   
4  126.41260       0.0       0.0  ...        0.0         0.0         0.0   

   6~7_takeoff  7~8_takeoff  8~9_takeoff  9~10_takeoff  10~11_takeoff  \
0          0.0          0.0          0.0           0.0            0.0   
1          0.0          0.0          0.0           0.0            0.0   
2          0.0          0.0          0.0           0.0            0.0   
3          0.0          0.0          0.0           0.0            0.0   
4          0.0          0.0          0.0           1.0            0.0   

   11~12_takeoff  18~20_ride  
0            0.0         0.0  
1            0.0         5.0  
2            0.0         2.0  
3            0.0        53.0  
4            0.0         0.0  

[5 rows x 21 columns]

In [5]:
data.tail()

id        date  bus_route_id in_out  station_code  station_name  \
415418  415418  2019-09-30      32820000     시내          1129  한림환승정류장(한림리)   
415419  415419  2019-09-30      32820000     시내          1564     제주시외버스터미널   
415420  415420  2019-09-30      32820000     시내          2322          해병부대   
415421  415421  2019-09-30      32820000     시내          3291  애월환승정류장(애월리)   
415422  415422  2019-09-30      32820000     시내       6115100    서귀포시외버스터미널   

        latitude  longitude  6~7_ride  7~8_ride  ...  9~10_ride  10~11_ride  \
415418  33.41437  126.26336       4.0       0.0  ...        0.0         0.0   
415419  33.49946  126.51479       4.0       0.0  ...        0.0         0.0   
415420  33.23100  126.26273       0.0       0.0  ...        0.0         0.0   
415421  33.46483  126.31870       1.0       0.0  ...        0.0         0.0   
415422  33.24873  126.50799       0.0       0.0  ...        0.0         0.0   

        11~12_ride  6~7_takeoff  7~8_takeoff  8~9_takeoff  9~10_takeoff  \
415418         0.0          0.0          0.0          0.0           0.0   
415419         0.0          0.0          0.0          0.0           0.0   
415420         0.0          0.0          1.0          0.0           0.0   
415421         0.0          0.0          0.0          0.0           0.0   
415422         0.0          0.0          0.0          4.0           0.0   

        10~11_takeoff  11~12_takeoff  18~20_ride  
415418            0.0            0.0         0.0  
415419            0.0            0.0         0.0  
415420            0.0            0.0         0.0  
415421            0.0            0.0         0.0  
415422            0.0            0.0         0.0  

[5 rows x 21 columns]

data의 column  
- id : obs 역할
- date : 날짜 (2019-09-01 ~ 2019-09-30)
- bus_route_id : 버스 노선 id
- in_out : 시내/시외
- station_code : 승하차 정류소 id
- station_name : 승하차 정류소 이름
- latitude : 정류장 위도
- longitude : 정류장 경도
- 6 / 7 / 8 / 9 / 10 / 11 ride : 출근시간에 승차한 인원 수
- 6 / 7 / 8 / 9 / 10 / 11 takeoff : 출근시간에 하차한 인원 수
- target 18시-20시 ride : 퇴근시간에 승차한 인원 수

In [6]:
data.describe()

id  bus_route_id  station_code       latitude  \
count  415423.000000  4.154230e+05  4.154230e+05  415423.000000   
mean   207711.000000  2.691935e+07  4.209311e+04      33.420375   
std    119922.434776  3.924652e+06  4.971504e+05       0.107996   
min         0.000000  4.270000e+06  1.000000e+00      33.208350   
25%    103855.500000  2.346000e+07  3.200000e+02      33.291080   
50%    207711.000000  2.803000e+07  1.130000e+03      33.478850   
75%    311566.500000  3.003000e+07  2.226000e+03      33.500020   
max    415422.000000  3.282000e+07  6.115101e+06      33.963640   

           longitude       6~7_ride       7~8_ride       8~9_ride  \
count  415423.000000  415423.000000  415423.000000  415423.000000   
mean      126.533574       0.305893       0.829699       0.815350   
std         0.140986       1.109766       2.255116       2.317561   
min       126.165040       0.000000       0.000000       0.000000   
25%       126.475780       0.000000       0.000000       0.000000   
50%       126.529770       0.000000       0.000000       0.000000   
75%       126.575890       0.000000       1.000000       1.000000   
max       126.965670      85.000000      94.000000     136.000000   

           9~10_ride     10~11_ride     11~12_ride    6~7_takeoff  \
count  415423.000000  415423.000000  415423.000000  415423.000000   
mean        0.642475       0.599618       0.579393       0.112870   
std         1.959844       1.885941       1.942137       0.597714   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         1.000000       0.000000       0.000000       0.000000   
max        78.000000     124.000000      99.000000      45.000000   

         7~8_takeoff    8~9_takeoff   9~10_takeoff  10~11_takeoff  \
count  415423.000000  415423.000000  415423.000000  415423.000000   
mean        0.344870       0.516481       0.430922       0.408001   
std         1.279179       1.658850       1.485124       1.412839   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max        66.000000      59.000000      65.000000      52.000000   

       11~12_takeoff     18~20_ride  
count  415423.000000  415423.000000  
mean        0.402874       1.242095  
std         1.446080       4.722287  
min         0.000000       0.000000  
25%         0.000000       0.000000  
50%         0.000000       0.000000  
75%         0.000000       1.000000  
max        81.000000     272.000000

### data 특정 조건 탐색

#### 만약 정류장이 제주썬호텔일때

In [7]:
# 제주썬호텔 정류장 코드 344
data[data['station_code']==344].iloc[:,:14]

id        date  bus_route_id in_out  station_code station_name  \
0            0  2019-09-01       4270000     시외           344        제주썬호텔   
11538    11538  2019-09-02       4270000     시외           344        제주썬호텔   
26498    26498  2019-09-03       4270000     시외           344        제주썬호텔   
41983    41983  2019-09-04       4270000     시외           344        제주썬호텔   
56722    56722  2019-09-05       4270000     시외           344        제주썬호텔   
72428    72428  2019-09-06       4270000     시외           344        제주썬호텔   
87780    87780  2019-09-07       4270000     시외           344        제주썬호텔   
97978    97978  2019-09-08       4270000     시외           344        제주썬호텔   
109560  109560  2019-09-09       4270000     시외           344        제주썬호텔   
124857  124857  2019-09-10       4270000     시외           344        제주썬호텔   
140862  140862  2019-09-11       4270000     시외           344        제주썬호텔   
156277  156277  2019-09-12       4270000     시외           344        제주썬호텔   
167652  167652  2019-09-13       4270000     시외           344        제주썬호텔   
176360  176360  2019-09-14       4270000     시외           344        제주썬호텔   
186996  186996  2019-09-15       4270000     시외           344        제주썬호텔   
198205  198205  2019-09-16       4270000     시외           344        제주썬호텔   
213892  213892  2019-09-17       4270000     시외           344        제주썬호텔   
229852  229852  2019-09-18       4270000     시외           344        제주썬호텔   
245619  245619  2019-09-19       4270000     시외           344        제주썬호텔   
261649  261649  2019-09-20       4270000     시외           344        제주썬호텔   
277476  277476  2019-09-21       4270000     시외           344        제주썬호텔   
289010  289010  2019-09-22       4270000     시외           344        제주썬호텔   
295063  295063  2019-09-23       4270000     시외           344        제주썬호텔   
310687  310687  2019-09-24       4270000     시외           344        제주썬호텔   
326655  326655  2019-09-25       4270000     시외           344        제주썬호텔   
342518  342518  2019-09-26       4270000     시외           344        제주썬호텔   
358388  358388  2019-09-27       4270000     시외           344        제주썬호텔   
374196  374196  2019-09-28       4270000     시외           344        제주썬호텔   
387415  387415  2019-09-29       4270000     시외           344        제주썬호텔   
399539  399539  2019-09-30       4270000     시외           344        제주썬호텔   

        latitude  longitude  6~7_ride  7~8_ride  8~9_ride  9~10_ride  \
0        33.4899  126.49373       0.0       1.0       2.0        5.0   
11538    33.4899  126.49373       1.0       8.0       3.0        3.0   
26498    33.4899  126.49373       2.0       4.0       2.0        7.0   
41983    33.4899  126.49373       3.0       3.0       1.0        2.0   
56722    33.4899  126.49373       2.0       4.0       8.0        0.0   
72428    33.4899  126.49373       0.0       3.0       3.0        2.0   
87780    33.4899  126.49373       1.0       1.0       2.0        0.0   
97978    33.4899  126.49373       1.0       1.0       3.0        2.0   
109560   33.4899  126.49373       4.0       3.0       3.0        0.0   
124857   33.4899  126.49373       1.0       2.0       1.0        1.0   
140862   33.4899  126.49373       2.0       5.0       3.0        0.0   
156277   33.4899  126.49373       1.0       1.0       3.0        3.0   
167652   33.4899  126.49373       0.0       1.0       3.0        1.0   
176360   33.4899  126.49373       0.0       1.0       2.0        2.0   
186996   33.4899  126.49373       0.0       1.0       3.0        0.0   
198205   33.4899  126.49373       3.0       4.0       3.0        1.0   
213892   33.4899  126.49373       1.0       2.0       4.0        2.0   
229852   33.4899  126.49373       3.0       1.0       4.0        7.0   
245619   33.4899  126.49373       3.0       4.0       2.0        2.0   
261649   33.4899  126.49373       1.0       3.0       4.0        2.0   
277476   33.4899  126.49373       0.0       3.0       4.0        0.0   
289010   

정류소 id가 344일때 추출된 행의 개수 30개  
9월 1일 - 9월 30일까지의 데이터

In [8]:
# 제주썬호텔 정류장 코드 343
data[data['station_code']==343].iloc[:,:14]

id        date  bus_route_id in_out  station_code station_name  \
25          25  2019-09-01       4280000     시외           343        제주썬호텔   
11564    11564  2019-09-02       4280000     시외           343        제주썬호텔   
26526    26526  2019-09-03       4280000     시외           343        제주썬호텔   
42010    42010  2019-09-04       4280000     시외           343        제주썬호텔   
56747    56747  2019-09-05       4280000     시외           343        제주썬호텔   
72452    72452  2019-09-06       4280000     시외           343        제주썬호텔   
98007    98007  2019-09-08       4280000     시외           343        제주썬호텔   
109587  109587  2019-09-09       4280000     시외           343        제주썬호텔   
124884  124884  2019-09-10       4280000     시외           343        제주썬호텔   
140885  140885  2019-09-11       4280000     시외           343        제주썬호텔   
176387  176387  2019-09-14       4280000     시외           343        제주썬호텔   
187023  187023  2019-09-15       4280000     시외           343        제주썬호텔   
198227  198227  2019-09-16       4280000     시외           343        제주썬호텔   
213922  213922  2019-09-17       4280000     시외           343        제주썬호텔   
229878  229878  2019-09-18       4280000     시외           343        제주썬호텔   
245646  245646  2019-09-19       4280000     시외           343        제주썬호텔   
261679  261679  2019-09-20       4280000     시외           343        제주썬호텔   
289030  289030  2019-09-22       4280000     시외           343        제주썬호텔   
295090  295090  2019-09-23       4280000     시외           343        제주썬호텔   
310712  310712  2019-09-24       4280000     시외           343        제주썬호텔   
326680  326680  2019-09-25       4280000     시외           343        제주썬호텔   
342545  342545  2019-09-26       4280000     시외           343        제주썬호텔   
358413  358413  2019-09-27       4280000     시외           343        제주썬호텔   
374224  374224  2019-09-28       4280000     시외           343        제주썬호텔   
387442  387442  2019-09-29       4280000     시외           343        제주썬호텔   
399569  399569  2019-09-30       4280000     시외           343        제주썬호텔   

        latitude  longitude  6~7_ride  7~8_ride  8~9_ride  9~10_ride  \
25      33.48972   126.4937       0.0       1.0       0.0        0.0   
11564   33.48972   126.4937       0.0       0.0       0.0        0.0   
26526   33.48972   126.4937       0.0       0.0       0.0        3.0   
42010   33.48972   126.4937       0.0       0.0       0.0        0.0   
56747   33.48972   126.4937       0.0       0.0       0.0        1.0   
72452   33.48972   126.4937       0.0       1.0       0.0        1.0   
98007   33.48972   126.4937       0.0       0.0       0.0        4.0   
109587  33.48972   126.4937       0.0       0.0       0.0        0.0   
124884  33.48972   126.4937       0.0       0.0       0.0        0.0   
140885  33.48972   126.4937       0.0       0.0       0.0        0.0   
176387  33.48972   126.4937       0.0       0.0       0.0        1.0   
187023  33.48972   126.4937       0.0       0.0       0.0        0.0   
198227  33.48972   126.4937       0.0       0.0       0.0        0.0   
213922  33.48972   126.4937       0.0       0.0       0.0        0.0   
229878  33.48972   126.4937       0.0       0.0       0.0        0.0   
245646  33.48972   126.4937       0.0       0.0       0.0        2.0   
261679  33.48972   126.4937       0.0       0.0       0.0        0.0   
289030  33.48972   126.4937       0.0       1.0       0.0        0.0   
295090  33.48972   126.4937       0.0       0.0       0.0        0.0   
310712  33.48972   126.4937       0.0       0.0       1.0        0.0   
326680  33.48972   126.4937       0.0       0.0       0.0        0.0   
342545  33.48972   126.4937       0.0       0.0       0.0        0.0   
358413  33.48972   126.4937       0.0       0.0       0.0        0.0   
374224  33.48972   126.4937       0.0       0.0       0.0        0.0   
387442  33.48972   126.4937       0.0       0.0       0.0        0.0   
399569  33.48972   126.4937      

정류소 id가 343일때 추출된 행의 개수 26개  
9월 1일 - 9월 30일 중에서 7일, 12일, 13일, 21일 없음

#### 승하차인원이 모두 0명인 경우

In [9]:
data[(data.iloc[:,8:14].sum(axis=1)==0) & (data.iloc[:,14:20].sum(axis=1)==0)]

Empty DataFrame
Columns: [id, date, bus_route_id, in_out, station_code, station_name, latitude, longitude, 6~7_ride, 7~8_ride, 8~9_ride, 9~10_ride, 10~11_ride, 11~12_ride, 6~7_takeoff, 7~8_takeoff, 8~9_takeoff, 9~10_takeoff, 10~11_takeoff, 11~12_takeoff, 18~20_ride]
Index: []

[0 rows x 21 columns]

출근시간에 승차인원수가 0명이고 동시에 하차인원수가 0명인 행은 없음

#### station_code가 344일때 요일별로 확인하기

##### 월요일

In [10]:
data[((data['date']=='2019-09-02') | (data['date']=='2019-09-09') | (data['date']=='2019-09-16') | (data['date']=='2019-09-23') | (data['date']=='2019-09-30')) & (data['station_code']==344)]

id        date  bus_route_id in_out  station_code station_name  \
11538    11538  2019-09-02       4270000     시외           344        제주썬호텔   
109560  109560  2019-09-09       4270000     시외           344        제주썬호텔   
198205  198205  2019-09-16       4270000     시외           344        제주썬호텔   
295063  295063  2019-09-23       4270000     시외           344        제주썬호텔   
399539  399539  2019-09-30       4270000     시외           344        제주썬호텔   

        latitude  longitude  6~7_ride  7~8_ride  ...  9~10_ride  10~11_ride  \
11538    33.4899  126.49373       1.0       8.0  ...        3.0         0.0   
109560   33.4899  126.49373       4.0       3.0  ...        0.0         2.0   
198205   33.4899  126.49373       3.0       4.0  ...        1.0         0.0   
295063   33.4899  126.49373       5.0       5.0  ...        3.0         0.0   
399539   33.4899  126.49373       3.0       5.0  ...       10.0         4.0   

        11~12_ride  6~7_takeoff  7~8_takeoff  8~9_takeoff  9~10_takeoff  \
11538          4.0          0.0          0.0          0.0           0.0   
109560         1.0          0.0          0.0          0.0           0.0   
198205         3.0          0.0          0.0          0.0           0.0   
295063         3.0          0.0          0.0          0.0           0.0   
399539         2.0          0.0          3.0          0.0           1.0   

        10~11_takeoff  11~12_takeoff  18~20_ride  
11538             0.0            0.0         1.0  
109560            0.0            0.0         0.0  
198205            0.0            0.0         5.0  
295063            0.0            0.0         1.0  
399539            0.0            0.0         3.0  

[5 rows x 21 columns]

##### 화요일

In [11]:
data[((data['date']=='2019-09-03') | (data['date']=='2019-09-10') | (data['date']=='2019-09-17') | (data['date']=='2019-09-24')) & (data['station_code']==344)]

id        date  bus_route_id in_out  station_code station_name  \
26498    26498  2019-09-03       4270000     시외           344        제주썬호텔   
124857  124857  2019-09-10       4270000     시외           344        제주썬호텔   
213892  213892  2019-09-17       4270000     시외           344        제주썬호텔   
310687  310687  2019-09-24       4270000     시외           344        제주썬호텔   

        latitude  longitude  6~7_ride  7~8_ride  ...  9~10_ride  10~11_ride  \
26498    33.4899  126.49373       2.0       4.0  ...        7.0         0.0   
124857   33.4899  126.49373       1.0       2.0  ...        1.0         3.0   
213892   33.4899  126.49373       1.0       2.0  ...        2.0         0.0   
310687   33.4899  126.49373       1.0       4.0  ...        3.0         0.0   

        11~12_ride  6~7_takeoff  7~8_takeoff  8~9_takeoff  9~10_takeoff  \
26498          4.0          0.0          0.0          0.0           0.0   
124857         0.0          0.0          2.0          0.0           0.0   
213892         2.0          0.0          0.0          0.0           0.0   
310687         1.0          0.0          0.0          0.0           1.0   

        10~11_takeoff  11~12_takeoff  18~20_ride  
26498             0.0            0.0         0.0  
124857            0.0            0.0         0.0  
213892            0.0            0.0         0.0  
310687            0.0            0.0         1.0  

[4 rows x 21 columns]

##### 수요일

In [12]:
data[((data['date']=='2019-09-04') | (data['date']=='2019-09-11') | (data['date']=='2019-09-18') | (data['date']=='2019-09-25')) & (data['station_code']==344)]

id        date  bus_route_id in_out  station_code station_name  \
41983    41983  2019-09-04       4270000     시외           344        제주썬호텔   
140862  140862  2019-09-11       4270000     시외           344        제주썬호텔   
229852  229852  2019-09-18       4270000     시외           344        제주썬호텔   
326655  326655  2019-09-25       4270000     시외           344        제주썬호텔   

        latitude  longitude  6~7_ride  7~8_ride  ...  9~10_ride  10~11_ride  \
41983    33.4899  126.49373       3.0       3.0  ...        2.0         8.0   
140862   33.4899  126.49373       2.0       5.0  ...        0.0         2.0   
229852   33.4899  126.49373       3.0       1.0  ...        7.0         3.0   
326655   33.4899  126.49373       5.0       3.0  ...        1.0         1.0   

        11~12_ride  6~7_takeoff  7~8_takeoff  8~9_takeoff  9~10_takeoff  \
41983          0.0          0.0          0.0          0.0           0.0   
140862         5.0          0.0          0.0          0.0           0.0   
229852         6.0          0.0          0.0          0.0           0.0   
326655         0.0          0.0          0.0          0.0           0.0   

        10~11_takeoff  11~12_takeoff  18~20_ride  
41983             0.0            0.0         0.0  
140862            0.0            0.0         5.0  
229852            0.0            0.0         2.0  
326655            0.0            0.0         1.0  

[4 rows x 21 columns]

##### 목요일

In [13]:
data[((data['date']=='2019-09-05') | (data['date']=='2019-09-12') | (data['date']=='2019-09-19') | (data['date']=='2019-09-26')) & (data['station_code']==344)]

id        date  bus_route_id in_out  station_code station_name  \
56722    56722  2019-09-05       4270000     시외           344        제주썬호텔   
156277  156277  2019-09-12       4270000     시외           344        제주썬호텔   
245619  245619  2019-09-19       4270000     시외           344        제주썬호텔   
342518  342518  2019-09-26       4270000     시외           344        제주썬호텔   

        latitude  longitude  6~7_ride  7~8_ride  ...  9~10_ride  10~11_ride  \
56722    33.4899  126.49373       2.0       4.0  ...        0.0         3.0   
156277   33.4899  126.49373       1.0       1.0  ...        3.0         4.0   
245619   33.4899  126.49373       3.0       4.0  ...        2.0         1.0   
342518   33.4899  126.49373       3.0       1.0  ...        2.0         2.0   

        11~12_ride  6~7_takeoff  7~8_takeoff  8~9_takeoff  9~10_takeoff  \
56722          4.0          0.0          0.0          0.0           0.0   
156277         0.0          0.0          0.0          0.0           0.0   
245619         7.0          0.0          0.0          0.0           0.0   
342518         0.0          0.0          0.0          0.0           0.0   

        10~11_takeoff  11~12_takeoff  18~20_ride  
56722             0.0            0.0         3.0  
156277            0.0            0.0         4.0  
245619            0.0            0.0         3.0  
342518            0.0            0.0         2.0  

[4 rows x 21 columns]

##### 금요일

In [14]:
data[((data['date']=='2019-09-06') | (data['date']=='2019-09-13') | (data['date']=='2019-09-20') | (data['date']=='2019-09-27')) & (data['station_code']==344)]

id        date  bus_route_id in_out  station_code station_name  \
72428    72428  2019-09-06       4270000     시외           344        제주썬호텔   
167652  167652  2019-09-13       4270000     시외           344        제주썬호텔   
261649  261649  2019-09-20       4270000     시외           344        제주썬호텔   
358388  358388  2019-09-27       4270000     시외           344        제주썬호텔   

        latitude  longitude  6~7_ride  7~8_ride  ...  9~10_ride  10~11_ride  \
72428    33.4899  126.49373       0.0       3.0  ...        2.0         5.0   
167652   33.4899  126.49373       0.0       1.0  ...        1.0         4.0   
261649   33.4899  126.49373       1.0       3.0  ...        2.0         0.0   
358388   33.4899  126.49373       2.0       4.0  ...        0.0         5.0   

        11~12_ride  6~7_takeoff  7~8_takeoff  8~9_takeoff  9~10_takeoff  \
72428          2.0          0.0          0.0          0.0           0.0   
167652         0.0          0.0          0.0          0.0           0.0   
261649         3.0          0.0          0.0          0.0           0.0   
358388         0.0          0.0          0.0          0.0           0.0   

        10~11_takeoff  11~12_takeoff  18~20_ride  
72428             0.0            0.0         0.0  
167652            0.0            0.0         3.0  
261649            4.0            0.0         2.0  
358388            0.0            0.0         1.0  

[4 rows x 21 columns]

##### 토요일

In [15]:
data[((data['date']=='2019-09-07') | (data['date']=='2019-09-14') | (data['date']=='2019-09-21') | (data['date']=='2019-09-28')) & (data['station_code']==344)]

id        date  bus_route_id in_out  station_code station_name  \
87780    87780  2019-09-07       4270000     시외           344        제주썬호텔   
176360  176360  2019-09-14       4270000     시외           344        제주썬호텔   
277476  277476  2019-09-21       4270000     시외           344        제주썬호텔   
374196  374196  2019-09-28       4270000     시외           344        제주썬호텔   

        latitude  longitude  6~7_ride  7~8_ride  ...  9~10_ride  10~11_ride  \
87780    33.4899  126.49373       1.0       1.0  ...        0.0         0.0   
176360   33.4899  126.49373       0.0       1.0  ...        2.0         0.0   
277476   33.4899  126.49373       0.0       3.0  ...        0.0         0.0   
374196   33.4899  126.49373       0.0       1.0  ...        9.0        13.0   

        11~12_ride  6~7_takeoff  7~8_takeoff  8~9_takeoff  9~10_takeoff  \
87780          0.0          0.0          0.0          0.0           0.0   
176360         3.0          0.0          0.0          0.0           0.0   
277476         3.0          0.0          0.0          0.0           0.0   
374196         0.0          0.0          0.0          0.0           0.0   

        10~11_takeoff  11~12_takeoff  18~20_ride  
87780             0.0            0.0         2.0  
176360            0.0            4.0         1.0  
277476            0.0            1.0         0.0  
374196            0.0            0.0         3.0  

[4 rows x 21 columns]

##### 일요일

In [16]:
data[((data['date']=='2019-09-01') | (data['date']=='2019-09-08') | (data['date']=='2019-09-15') | (data['date']=='2019-09-22') | (data['date']=='2019-09-29')) & (data['station_code']==344)]

id        date  bus_route_id in_out  station_code station_name  \
0            0  2019-09-01       4270000     시외           344        제주썬호텔   
97978    97978  2019-09-08       4270000     시외           344        제주썬호텔   
186996  186996  2019-09-15       4270000     시외           344        제주썬호텔   
289010  289010  2019-09-22       4270000     시외           344        제주썬호텔   
387415  387415  2019-09-29       4270000     시외           344        제주썬호텔   

        latitude  longitude  6~7_ride  7~8_ride  ...  9~10_ride  10~11_ride  \
0        33.4899  126.49373       0.0       1.0  ...        5.0         2.0   
97978    33.4899  126.49373       1.0       1.0  ...        2.0         0.0   
186996   33.4899  126.49373       0.0       1.0  ...        0.0         3.0   
289010   33.4899  126.49373       0.0       1.0  ...        0.0         1.0   
387415   33.4899  126.49373       1.0       2.0  ...        3.0         4.0   

        11~12_ride  6~7_takeoff  7~8_takeoff  8~9_takeoff  9~10_takeoff  \
0              6.0          0.0          0.0          0.0           0.0   
97978          0.0          0.0          0.0          0.0           0.0   
186996         2.0          0.0          0.0          0.0           1.0   
289010         0.0          0.0          0.0          0.0           0.0   
387415         2.0          0.0          0.0          0.0           0.0   

        10~11_takeoff  11~12_takeoff  18~20_ride  
0                 0.0            0.0         0.0  
97978             0.0            0.0         0.0  
186996            0.0            0.0         0.0  
289010            0.0            0.0         1.0  
387415            0.0            0.0         0.0  

[5 rows x 21 columns]

**요일변수를 추가해야할지 고민해보기**

#### 시간대별 승하차인원 확인하기

In [17]:
data.iloc[:, 8:20].sum()

6~7_ride         127075.0
7~8_ride         344676.0
8~9_ride         338715.0
9~10_ride        266899.0
10~11_ride       249095.0
11~12_ride       240693.0
6~7_takeoff       46889.0
7~8_takeoff      143267.0
8~9_takeoff      214558.0
9~10_takeoff     179015.0
10~11_takeoff    169493.0
11~12_takeoff    167363.0
dtype: float64

7-8시에 승차인원이 가장 많고 8-9시 하차인원이 가장 많음

In [18]:
# 총 승차인원, 하차인원 변수 추가
data['ride_total'] = data.iloc[:,8:14].sum(axis=1)
data['takeoff_total'] = data.iloc[:,14:20].sum(axis=1)

data.head()

id        date  bus_route_id in_out  station_code  station_name  latitude  \
0   0  2019-09-01       4270000     시외           344         제주썬호텔  33.48990   
1   1  2019-09-01       4270000     시외           357          한라병원  33.48944   
2   2  2019-09-01       4270000     시외           432          정존마을  33.48181   
3   3  2019-09-01       4270000     시내          1579  제주국제공항(600번)  33.50577   
4   4  2019-09-01       4270000     시내          1646      중문관광단지입구  33.25579   

   longitude  6~7_ride  7~8_ride  ...  11~12_ride  6~7_takeoff  7~8_takeoff  \
0  126.49373       0.0       1.0  ...         6.0          0.0          0.0   
1  126.48508       1.0       4.0  ...         6.0          0.0          0.0   
2  126.47352       1.0       1.0  ...         0.0          0.0          0.0   
3  126.49252       0.0      17.0  ...        16.0          0.0          0.0   
4  126.41260       0.0       0.0  ...         0.0          0.0          0.0   

   8~9_takeoff  9~10_takeoff  10~11_takeoff  11~12_takeoff  18~20_ride  \
0          0.0           0.0            0.0            0.0         0.0   
1          0.0           0.0            0.0            0.0         5.0   
2          0.0           0.0            0.0            0.0         2.0   
3          0.0           0.0            0.0            0.0        53.0   
4          0.0           1.0            0.0            0.0         0.0   

   ride_total  takeoff_total  
0        16.0            0.0  
1        22.0            0.0  
2         4.0            0.0  
3        79.0            0.0  
4         0.0            1.0  

[5 rows x 23 columns]

In [19]:
# 승차인원이 많은 정류장
data.groupby(['station_code','station_name'])['ride_total','takeoff_total'].mean().sort_values(by='ride_total',ascending=False).head(10)

ride_total  takeoff_total
station_code station_name                                 
1579         제주국제공항(600번)         82.967742       0.483871
3351         제주국제공항(일주동로,516도로)   47.150000       1.050000
6115000      여객선대합실               36.620690       0.000000
1564         제주시외버스터미널            33.087622       0.617524
1919         비석거리                 24.075269       1.924731
604          삼양1동                 24.010638       0.148936
3487         첨단꿈에그린2단지아파트         21.435897       2.461538
1565         시외버스출발지(가상정류소)       21.038409       0.506173
3354         제주동여자중학교             20.000000       2.400000
3350         제주국제공항(평화로, 800번)    19.954545       0.602273

In [20]:
# 하차인원이 많은 정류장
data.groupby(['station_code','station_name'])['ride_total','takeoff_total'].mean().sort_values(by='takeoff_total',ascending=False).head(10)

ride_total  takeoff_total
station_code station_name                            
2972         제주국제공항(종점)       0.550505      33.257576
2674         성판악              2.901639      27.032787
1355         제주국제공항(신제주방면)    9.406250      20.041667
1090         한담동              3.754098      19.803279
2971         제주한라대학교(종점)     17.402439      19.292683
2969         제주버스터미널(종점)      0.122881      19.163842
638          제주국제공항(구제주방면)   13.747856      18.831904
2022         향토오일시장           6.000000      13.882353
3486         제주대학교병원          6.104348      13.382609
2970         제주시외버스터미널(종점)    0.188347      13.380759

In [21]:
data.groupby('station_code')['6~7_ride'].mean().sort_values(ascending=False)

station_code
1564    4.094576
3372    3.183333
1919    3.010753
1565    2.460905
604     2.372340
          ...   
2323    0.000000
2321    0.000000
2320    0.000000
2319    0.000000
1918    0.000000
Name: 6~7_ride, Length: 3563, dtype: float64

In [22]:
data.groupby('station_code')['7~8_ride'].mean().sort_values(ascending=False)

station_code
1579    14.000000
604      8.787234
3357     8.162162
2496     7.897959
3351     7.483333
          ...    
656      0.000000
655      0.000000
2960     0.000000
3108     0.000000
2642     0.000000
Name: 7~8_ride, Length: 3563, dtype: float64

In [23]:
data.groupby('station_code')['8~9_ride'].mean().sort_values(ascending=False)

station_code
1579       14.806452
4205       13.000000
3351       10.733333
3333       10.343750
6115000     9.482759
             ...    
890         0.000000
886         0.000000
883         0.000000
2900        0.000000
4111        0.000000
Name: 8~9_ride, Length: 3563, dtype: float64

In [24]:
data.groupby('station_code')['9~10_ride'].mean().sort_values(ascending=False)

station_code
1579       18.645161
6115000    11.448276
3351        8.516667
1564        5.791377
3350        4.920455
             ...    
4066        0.000000
3194        0.000000
2042        0.000000
3521        0.000000
3404        0.000000
Name: 9~10_ride, Length: 3563, dtype: float64

In [25]:
data.groupby('station_code')['10~11_ride'].mean().sort_values(ascending=False)

station_code
1579       16.258065
6115000    10.896552
3351        9.533333
1564        5.324061
3290        4.089655
             ...    
2099        0.000000
2097        0.000000
2860        0.000000
2096        0.000000
1402        0.000000
Name: 10~11_ride, Length: 3563, dtype: float64

In [26]:
data.groupby('station_code')['11~12_ride'].mean().sort_values(ascending=False)

station_code
1579       17.193548
3351       10.350000
6115038     6.500000
6115010     5.000000
6115047     4.869565
             ...    
4126        0.000000
3635        0.000000
3634        0.000000
1702        0.000000
1753        0.000000
Name: 11~12_ride, Length: 3563, dtype: float64

## 두번째 bus_bts

### bus_bts 기본정보

In [28]:
bus_bts.shape

(2409414, 13)

데이터 행 개수 : 2409414개  
데이터 열 개수 : 13개

In [29]:
bus_bts.head()

user_card_id  bus_route_id     vhc_id  geton_date geton_time  \
0  1.010010e+15      23000000  149793674  2019-09-10   06:34:45   
1  1.010010e+15      23000000  149793674  2019-09-10   06:34:58   
2  1.019160e+15      21420000  149793535  2019-09-10   07:19:07   
3  1.019150e+15      21420000  149793512  2019-09-09   09:14:47   
4  1.010010e+15      21420000  149793512  2019-09-09   09:28:53   

   geton_station_code geton_station_name getoff_date getoff_time  \
0                 360              노형오거리  2019-09-10    07:10:31   
1                 360              노형오거리  2019-09-10    06:56:27   
2                2495     동광환승정류장4(제주방면)  2019-09-10    07:40:29   
3                3282    대정환승정류장(대정읍사무소)  2019-09-09    10:02:46   
4                2820              삼정지에듀  2019-09-09    10:21:37   

   getoff_station_code getoff_station_name  user_category  user_count  
0                592.0              화북초등학교              1           1  
1               3273.0          고산동산(광양방면)              1           1  
2                431.0                정존마을              4           1  
3                431.0                정존마을              1           1  
4               2972.0          제주국제공항(종점)              4           1

In [30]:
bus_bts.tail()

user_card_id  bus_route_id     vhc_id  geton_date geton_time  \
2409409  6.573162e+15      30420000  149797565  2019-10-16   07:08:31   
2409410  9.441160e+15      30420000  149797565  2019-10-16   07:16:31   
2409411  9.446038e+15      25070000  149797565  2019-10-16   08:29:05   
2409412  4.309690e+15      25070000  149797565  2019-10-16   08:40:32   
2409413  5.272898e+15      25070000  149797565  2019-10-16   08:40:35   

         geton_station_code geton_station_name getoff_date getoff_time  \
2409409                1937                L마트         NaN         NaN   
2409410                1908              한라산교회         NaN         NaN   
2409411                1882               고도농원         NaN         NaN   
2409412                1938           주공아파트5단지         NaN         NaN   
2409413                1938           주공아파트5단지         NaN         NaN   

         getoff_station_code getoff_station_name  user_category  user_count  
2409409                  NaN                 NaN              1           1  
2409410                  NaN                 NaN              1           1  
2409411                  NaN                 NaN              1           1  
2409412                  NaN                 NaN              1           1  
2409413                  NaN                 NaN              1           1

bus_bts의 columns  
- user_card_id : 승객 카드 id
- bus_route_id : 버스 노선 id
- vhc_id : 버스 차량 id
- geton_date : 승차 날짜
- geton_time : 승차 시간
- geton_station_code : 승차 정류장 id
- geton_station_name : 승차 정류장 이름
- getoff_date : 하차 날짜
- getoff_time : 하차 시간
- getoff_station_code : 하차 정류장 id
- getoff_station_name : 하차 정류장 이름
- user_category : 승객 유형 구분
- user_count : 해당 버스카드로 계산한 인원수

In [31]:
bus_bts.describe()

user_card_id  bus_route_id        vhc_id  geton_station_code  \
count  2.409414e+06  2.409414e+06  2.409414e+06        2.409414e+06   
mean   2.920959e+15  2.545786e+07  1.497948e+08        1.197044e+05   
std    2.249502e+16  4.017593e+06  1.808600e+03        8.422198e+05   
min    1.355555e+07  4.270000e+06  1.497911e+08        0.000000e+00   
25%    1.014140e+15  2.313000e+07  1.497932e+08        2.780000e+02   
50%    1.019150e+15  2.400000e+07  1.497936e+08        6.030000e+02   
75%    5.188320e+15  2.870000e+07  1.497970e+08        2.057000e+03   
max    1.000700e+19  3.282000e+07  1.497989e+08        6.115101e+06   

       getoff_station_code  user_category    user_count  
count         1.513678e+06   2.409414e+06  2.409414e+06  
mean          8.406106e+04   4.073226e+00  1.012212e+00  
std           7.058707e+05   6.523765e+00  1.354882e-01  
min           1.000000e+00   1.000000e+00  1.000000e+00  
25%           2.860000e+02   1.000000e+00  1.000000e+00  
50%           8.510000e+02   1.000000e+00  1.000000e+00  
75%           2.374000e+03   6.000000e+00  1.000000e+00  
max           6.115101e+06   3.000000e+01  1.000000e+01

### bus_bts 특정 조건 탐색

In [32]:
np.unique(bus_bts['geton_date'])

array(['2019-09-01', '2019-09-02', '2019-09-03', '2019-09-04',
       '2019-09-05', '2019-09-06', '2019-09-07', '2019-09-08',
       '2019-09-09', '2019-09-10', '2019-09-11', '2019-09-12',
       '2019-09-13', '2019-09-14', '2019-09-15', '2019-09-16',
       '2019-09-17', '2019-09-18', '2019-09-19', '2019-09-20',
       '2019-09-21', '2019-09-22', '2019-09-23', '2019-09-24',
       '2019-09-25', '2019-09-26', '2019-09-27', '2019-09-28',
       '2019-09-29', '2019-09-30', '2019-10-01', '2019-10-02',
       '2019-10-03', '2019-10-04', '2019-10-05', '2019-10-06',
       '2019-10-07', '2019-10-08', '2019-10-09', '2019-10-10',
       '2019-10-11', '2019-10-12', '2019-10-13', '2019-10-14',
       '2019-10-15', '2019-10-16'], dtype=object)

9월 1일부터 10월 16일까지 데이터가 있음  => 첫번째 데이터와 맞추기 위해 10월 데이터 제거하기

In [33]:
np.unique(bus_bts['geton_time'])

array(['06:00:00', '06:00:01', '06:00:02', ..., '11:59:57', '11:59:58',
       '11:59:59'], dtype=object)

승차 시간은 출근시간으로 제한됨. 6시-12시

In [34]:
len(np.unique(bus_bts['user_card_id']))

326776

행의 개수는 2409414개 이지만, 승객 카드는 326776개 이므로 같은 승객이 규칙적으로 승차할것으로 추측됨 => 카드번호를 기준으로 일정한 패턴으로 승차하면 직장인으로 추정됨

## 현재까지 고려사항

- 2019년 9월의 이벤트
    + 6일, 7일 : 태풍 링링
    + 12일, 13일, 14일 : 추석 연휴
    + 21일 : 태풍 타파
- 승객 유형에서 어린이, 청소년을 제외할지 생각해보기
- bus_bts에서 9월 데이터만 남기자. (10월 데이터에선 얻어올 정보가 없을까?)
- 명목형으로 요일변수, 공휴일변수를 추가할지 생각해보기
- 카드번호를 기준으로 규칙적으로 출퇴근하는 직장인 탐색 가능할 것으로 예상
- data에서 시내, 시외 버스 변수는 어떻게 활용해야할까?
- 날씨 데이터가 필요할까?
- 여러가지 그래프 많이 그리기 (R program)

----------------------------------------------------------------------------------------------------

## 승객 유형

In [36]:
np.unique(bus_bts['user_category'])

array([ 1,  2,  4,  6, 27, 28, 29, 30], dtype=int64)

1 : 일반  
2 : 어린이  
4 : 청소년    
6 : 경로  
27 : 장애일반  
28 : 장애동반    
29 : 유공일반  
30 : 유공동반

In [37]:
ctg2 = list(bus_bts[bus_bts['user_category']==2].index.values)

In [38]:
# 어린이로 분류된 특정 승객의 버스 이용 데이터 뽑아보기
bus_bts[bus_bts['user_card_id']==bus_bts['user_card_id'][ctg2[0]]]

user_card_id  bus_route_id     vhc_id  geton_date geton_time  \
15       1.019150e+15      31350000  149793472  2019-09-10   07:59:39   
77484    1.019150e+15      22210000  149793015  2019-09-16   08:06:41   
141526   1.019150e+15      31860000  149793610  2019-09-18   07:59:42   
437446   1.019150e+15      31860000  149793621  2019-09-05   08:02:01   
529409   1.019150e+15      23610000  149797022  2019-09-17   07:59:33   
1067392  1.019150e+15      23610000  149797085  2019-09-06   07:57:46   
1365286  1.019150e+15      31350000  149793474  2019-09-20   07:57:44   
1434428  1.019150e+15      22530000  149793485  2019-09-23   07:57:21   
1499567  1.019150e+15      31350000  149793472  2019-09-30   07:59:56   
1712983  1.019150e+15      31350000  149793473  2019-10-01   08:02:00   
1811033  1.019150e+15      23420000  149793471  2019-10-07   08:11:22   
1873211  1.019150e+15      22530000  149793483  2019-10-08   08:01:49   
2027876  1.019150e+15      31860000  149793621  2019-10-10   07:59:34   
2197784  1.019150e+15      23450000  149797106  2019-10-11   08:09:09   
2298847  1.019150e+15      23420000  149793471  2019-10-15   08:13:40   
2355713  1.019150e+15      22530000  149793415  2019-10-16   08:18:45   

         geton_station_code geton_station_name getoff_date getoff_time  \
15                      131            아라동주민센터  2019-09-10    08:03:10   
77484                   131            아라동주민센터  2019-09-16    08:10:08   
141526                  131            아라동주민센터  2019-09-18    08:01:47   
437446                  131            아라동주민센터  2019-09-05    08:05:45   
529409                  131            아라동주민센터  2019-09-17    08:03:21   
1067392                 131            아라동주민센터         NaN         NaN   
1365286                 131            아라동주민센터  2019-09-20    08:00:55   
1434428                 131            아라동주민센터  2019-09-23    07:59:29   
1499567                 131            아라동주민센터  2019-09-30    08:03:26   
1712983                 131            아라동주민센터  2019-10-01    08:05:24   
1811033                 131            아라동주민센터  2019-10-07    08:14:49   
1873211                 131            아라동주민센터  2019-10-08    08:04:07   
2027876                 131            아라동주민센터  2019-10-10    08:03:20   
2197784                 131            아라동주민센터  2019-10-11    08:12:31   
2298847                 131            아라동주민센터  2019-10-15    08:17:01   
2355713                 131            아라동주민센터  2019-10-16    08:21:55   

         getoff_station_code getoff_station_name  user_category  user_count  
15                    3279.0     제주여자중고등학교(광양방면)              2           1  
77484                 3279.0     제주여자중고등학교(광양방면)              2           1  
141526                3279.0     제주여자중고등학교(광양방면)              2           1  
437446                3279.0     제주여자중고등학교(광양방면)              2           1  
529409                3279.0     제주여자중고등학교(광양방면)              2           1  
1067392                  NaN                 NaN              2           1  
1365286               3279.0     제주여자중고등학교(광양방면)              2           1  
1434428               3279.0     제주여자중고등학교(광양방면)              2           1  
1499567               3279.0     제주여자중고등학교(광양방면)              2           1  
1712983               3279.0     제주여자중고등학교(광양방면)              2           1  
1811033               3279.0     제주여자중고등학교(광양방면)              2           1  
1873211               3279.0     제주여자중고등학교(광양방면)              2           1  
2027876               3279.0     제주여자중고등학교(광양방면)              2           1  
2197784               3279.0     제주여자중고등학교(광양방면)              2           1  
2298847               3279.0     제주여자중고등학교(광양방면)              2           1  
2355713               3279.0     제주여자중고등학교(광양방면)              2           1

In [39]:
bus_bts[bus_bts['user_card_id']==bus_bts['user_card_id'][ctg2[1]]]

user_card_id  bus_route_id     vhc_id  geton_date geton_time  \
75       1.010010e+15      29640000  149793538  2019-09-10   07:29:54   
35248    1.010010e+15      29540000  149797008  2019-09-11   07:43:00   
88090    1.010010e+15      23250000  149797047  2019-09-16   07:43:04   
158236   1.010010e+15      23250000  149797061  2019-09-18   07:42:42   
206818   1.010010e+15      29640000  149793570  2019-09-27   07:30:20   
...               ...           ...        ...         ...        ...   
2157423  1.010010e+15      23250000  149797072  2019-10-07   07:41:58   
2201308  1.010010e+15      23400000  149793547  2019-10-11   09:22:54   
2287172  1.010010e+15      29640000  149793570  2019-10-15   07:30:04   
2292763  1.010010e+15      23250000  149797026  2019-10-15   07:42:08   
2382319  1.010010e+15      23250000  149797027  2019-10-16   07:39:26   

         geton_station_code geton_station_name getoff_date getoff_time  \
75                      288            명도암마을회관  2019-09-10    07:37:44   
35248                   263               회천입구         NaN         NaN   
88090                   239            봉개동주민센터         NaN         NaN   
158236                  239            봉개동주민센터         NaN         NaN   
206818                  288            명도암마을회관  2019-09-27    07:37:38   
...                     ...                ...         ...         ...   
2157423                 239            봉개동주민센터         NaN         NaN   
2201308                 288            명도암마을회관         NaN         NaN   
2287172                 288            명도암마을회관  2019-10-15    07:37:42   
2292763                 239            봉개동주민센터  2019-10-15    07:57:17   
2382319                 239            봉개동주민센터         NaN         NaN   

         getoff_station_code getoff_station_name  user_category  user_count  
75                     239.0             봉개동주민센터              2           1  
35248                    NaN                 NaN              2           1  
88090                    NaN                 NaN              2           1  
158236                   NaN                 NaN              2           1  
206818                 239.0             봉개동주민센터              2           1  
...                      ...                 ...            ...         ...  
2157423                  NaN                 NaN              2           1  
2201308                  NaN                 NaN              4           1  
2287172                239.0             봉개동주민센터              4           1  
2292763                243.0           화북주공아파트입구              4           1  
2382319                  NaN                 NaN              4           1  

[61 rows x 13 columns]

## Final Data 만들기

[변수 구성]  

date, bus_route_id, station_code, station_name, latitude, logitude,  
6-7 ride ... 11-12 ride, ride_total  
6-7 takeoff ... 11-12 takeoff, takeoff_total  
type_general_ride, type_others_ride, type_others_ride, type_others_ride, holiday, weather, bus_type

In [41]:
final_data = pd.DataFrame()
final_data['date'] = data['date']
final_data['bus_route_id'] = data['bus_route_id']
final_data['station_code'] = data['station_code']
final_data['station_name'] = data['station_name']
final_data['latitude'] = data['latitude']
final_data['longitude'] = data['longitude']
final_data['6~7_ride'] = data['6~7_ride']
final_data['7~8_ride'] = data['7~8_ride']
final_data['8~9_ride'] = data['8~9_ride']
final_data['9~10_ride'] = data['9~10_ride']
final_data['10~11_ride'] = data['10~11_ride']
final_data['11~12_ride'] = data['11~12_ride']
final_data['6~7_takeoff'] = data['6~7_takeoff']
final_data['7~8_takeoff'] = data['7~8_takeoff']
final_data['8~9_takeoff'] = data['8~9_takeoff']
final_data['9~10_takeoff'] = data['9~10_takeoff']
final_data['10~11_takeoff'] = data['10~11_takeoff']
final_data['11~12_takeoff'] = data['11~12_takeoff']
final_data['ride_total'] = data['ride_total']
final_data['takeoff_total'] = data['takeoff_total']
final_data.head()

date  bus_route_id  station_code  station_name  latitude  longitude  \
0  2019-09-01       4270000           344         제주썬호텔  33.48990  126.49373   
1  2019-09-01       4270000           357          한라병원  33.48944  126.48508   
2  2019-09-01       4270000           432          정존마을  33.48181  126.47352   
3  2019-09-01       4270000          1579  제주국제공항(600번)  33.50577  126.49252   
4  2019-09-01       4270000          1646      중문관광단지입구  33.25579  126.41260   

   6~7_ride  7~8_ride  8~9_ride  9~10_ride  10~11_ride  11~12_ride  \
0       0.0       1.0       2.0        5.0         2.0         6.0   
1       1.0       4.0       4.0        2.0         5.0         6.0   
2       1.0       1.0       0.0        2.0         0.0         0.0   
3       0.0      17.0       6.0       26.0        14.0        16.0   
4       0.0       0.0       0.0        0.0         0.0         0.0   

   6~7_takeoff  7~8_takeoff  8~9_takeoff  9~10_takeoff  10~11_takeoff  \
0          0.0          0.0          0.0           0.0            0.0   
1          0.0          0.0          0.0           0.0            0.0   
2          0.0          0.0          0.0           0.0            0.0   
3          0.0          0.0          0.0           0.0            0.0   
4          0.0          0.0          0.0           1.0            0.0   

   11~12_takeoff  ride_total  takeoff_total  
0            0.0        16.0            0.0  
1            0.0        22.0            0.0  
2            0.0         4.0            0.0  
3            0.0        79.0            0.0  
4            0.0         0.0            1.0

In [42]:
final_data.tail()

date  bus_route_id  station_code  station_name  latitude  \
415418  2019-09-30      32820000          1129  한림환승정류장(한림리)  33.41437   
415419  2019-09-30      32820000          1564     제주시외버스터미널  33.49946   
415420  2019-09-30      32820000          2322          해병부대  33.23100   
415421  2019-09-30      32820000          3291  애월환승정류장(애월리)  33.46483   
415422  2019-09-30      32820000       6115100    서귀포시외버스터미널  33.24873   

        longitude  6~7_ride  7~8_ride  8~9_ride  9~10_ride  10~11_ride  \
415418  126.26336       4.0       0.0       0.0        0.0         0.0   
415419  126.51479       4.0       0.0       0.0        0.0         0.0   
415420  126.26273       0.0       0.0       0.0        0.0         0.0   
415421  126.31870       1.0       0.0       0.0        0.0         0.0   
415422  126.50799       0.0       0.0       0.0        0.0         0.0   

        11~12_ride  6~7_takeoff  7~8_takeoff  8~9_takeoff  9~10_takeoff  \
415418         0.0          0.0          0.0          0.0           0.0   
415419         0.0          0.0          0.0          0.0           0.0   
415420         0.0          0.0          1.0          0.0           0.0   
415421         0.0          0.0          0.0          0.0           0.0   
415422         0.0          0.0          0.0          4.0           0.0   

        10~11_takeoff  11~12_takeoff  ride_total  takeoff_total  
415418            0.0            0.0         4.0            0.0  
415419            0.0            0.0         4.0            0.0  
415420            0.0            0.0         0.0            1.0  
415421            0.0            0.0         1.0            0.0  
415422            0.0            0.0         0.0            4.0

### 공휴일 변수 만들기

In [43]:
holiday_date = []
holiday_date = ['2019-09-01', '2019-09-07','2019-09-08','2019-09-12','2019-09-13','2019-09-14','2019-09-15','2019-09-21','2019-09-22','2019-09-28','2019-09-29']

final_data['holiday'] = np.nan

In [45]:
holiday = []

for i in range(len(final_data)):
    if final_data['date'][i] in holiday_date:
        holiday.append(0)
    else:
        holiday.append(1)

In [46]:
final_data['holiday'] = holiday
final_data.head()

date  bus_route_id  station_code  station_name  latitude  longitude  \
0  2019-09-01       4270000           344         제주썬호텔  33.48990  126.49373   
1  2019-09-01       4270000           357          한라병원  33.48944  126.48508   
2  2019-09-01       4270000           432          정존마을  33.48181  126.47352   
3  2019-09-01       4270000          1579  제주국제공항(600번)  33.50577  126.49252   
4  2019-09-01       4270000          1646      중문관광단지입구  33.25579  126.41260   

   6~7_ride  7~8_ride  8~9_ride  9~10_ride  ...  11~12_ride  6~7_takeoff  \
0       0.0       1.0       2.0        5.0  ...         6.0          0.0   
1       1.0       4.0       4.0        2.0  ...         6.0          0.0   
2       1.0       1.0       0.0        2.0  ...         0.0          0.0   
3       0.0      17.0       6.0       26.0  ...        16.0          0.0   
4       0.0       0.0       0.0        0.0  ...         0.0          0.0   

   7~8_takeoff  8~9_takeoff  9~10_takeoff  10~11_takeoff  11~12_takeoff  \
0          0.0          0.0           0.0            0.0            0.0   
1          0.0          0.0           0.0            0.0            0.0   
2          0.0          0.0           0.0            0.0            0.0   
3          0.0          0.0           0.0            0.0            0.0   
4          0.0          0.0           1.0            0.0            0.0   

   ride_total  takeoff_total  holiday  
0        16.0            0.0        0  
1        22.0            0.0        0  
2         4.0            0.0        0  
3        79.0            0.0        0  
4         0.0            1.0        0  

[5 rows x 21 columns]

In [47]:
final_data.to_csv('final_data.csv')

## 승객 유형별 승하차인원 변수 추가한 데이터 확인하기
**(made from  R)**

일반 승객과 청소년 승객의 승하차인원 = type_general_geton / type_general_getoff  
그 외의 나머지 승객의 승하차인원 = type_others_geton / type_others_getoff

In [48]:
final_category = pd.read_csv('final_category.csv', encoding='ANSI')
final_category.head()

Unnamed: 0  X        date  bus_route_id  station_code  station_name  \
0           1  0  2019-09-01       4270000           344         제주썬호텔   
1           2  1  2019-09-01       4270000           357          한라병원   
2           3  2  2019-09-01       4270000           432          정존마을   
3           4  3  2019-09-01       4270000          1579  제주국제공항(600번)   
4           5  4  2019-09-01       4270000          1646      중문관광단지입구   

   latitude  longitude  X6.7_ride  X7.8_ride  ...  X9.10_takeoff  \
0  33.48990  126.49373          0          1  ...              0   
1  33.48944  126.48508          1          4  ...              0   
2  33.48181  126.47352          1          1  ...              0   
3  33.50577  126.49252          0         17  ...              0   
4  33.25579  126.41260          0          0  ...              1   

   X10.11_takeoff  X11.12_takeoff  ride_total  takeoff_total  holiday  \
0               0               0          16              0        0   
1               0               0          22              0        0   
2               0               0           4              0        0   
3               0               0          79              0        0   
4               0               0           0              1        0   

   type_general_geton  type_other_geton  type_general_getoff  \
0                15.0               1.0                  NaN   
1                22.0               0.0                  NaN   
2                 4.0               0.0                  NaN   
3                79.0               0.0                  NaN   
4                 NaN               NaN                  1.0   

   type_other_getoff  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                0.0  

[5 rows x 27 columns]

In [49]:
final_category['type_general_geton'] = final_category['type_general_geton'].fillna(0)
final_category['type_other_geton'] = final_category['type_other_geton'].fillna(0)
final_category.head()

Unnamed: 0  X        date  bus_route_id  station_code  station_name  \
0           1  0  2019-09-01       4270000           344         제주썬호텔   
1           2  1  2019-09-01       4270000           357          한라병원   
2           3  2  2019-09-01       4270000           432          정존마을   
3           4  3  2019-09-01       4270000          1579  제주국제공항(600번)   
4           5  4  2019-09-01       4270000          1646      중문관광단지입구   

   latitude  longitude  X6.7_ride  X7.8_ride  ...  X9.10_takeoff  \
0  33.48990  126.49373          0          1  ...              0   
1  33.48944  126.48508          1          4  ...              0   
2  33.48181  126.47352          1          1  ...              0   
3  33.50577  126.49252          0         17  ...              0   
4  33.25579  126.41260          0          0  ...              1   

   X10.11_takeoff  X11.12_takeoff  ride_total  takeoff_total  holiday  \
0               0               0          16              0        0   
1               0               0          22              0        0   
2               0               0           4              0        0   
3               0               0          79              0        0   
4               0               0           0              1        0   

   type_general_geton  type_other_geton  type_general_getoff  \
0                15.0               1.0                  NaN   
1                22.0               0.0                  NaN   
2                 4.0               0.0                  NaN   
3                79.0               0.0                  NaN   
4                 0.0               0.0                  1.0   

   type_other_getoff  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                0.0  

[5 rows x 27 columns]

In [50]:
final_category[final_category['type_other_geton'] + final_category['type_general_geton'] != final_category['ride_total']]

Empty DataFrame
Columns: [Unnamed: 0, X, date, bus_route_id, station_code, station_name, latitude, longitude, X6.7_ride, X7.8_ride, X8.9_ride, X9.10_ride, X10.11_ride, X11.12_ride, X6.7_takeoff, X7.8_takeoff, X8.9_takeoff, X9.10_takeoff, X10.11_takeoff, X11.12_takeoff, ride_total, takeoff_total, holiday, type_general_geton, type_other_geton, type_general_getoff, type_other_getoff]
Index: []

[0 rows x 27 columns]

In [51]:
final_category[final_category['type_other_getoff'] + final_category['type_general_getoff'] != final_category['takeoff_total']]

Unnamed: 0       X        date  bus_route_id  station_code  \
0                1       0  2019-09-01       4270000           344   
1                2       1  2019-09-01       4270000           357   
2                3       2  2019-09-01       4270000           432   
3                4       3  2019-09-01       4270000          1579   
9               10       9  2019-09-01       4270000          2008   
...            ...     ...         ...           ...           ...   
415412      415413  415412  2019-09-30      32810000          2468   
415414      415415  415414  2019-09-30      32820000            96   
415418      415419  415418  2019-09-30      32820000          1129   
415419      415420  415419  2019-09-30      32820000          1564   
415421      415422  415421  2019-09-30      32820000          3291   

         station_name  latitude  longitude  X6.7_ride  X7.8_ride  ...  \
0               제주썬호텔  33.48990  126.49373          0          1  ...   
1                한라병원  33.48944  126.48508          1          4  ...   
2                정존마을  33.48181  126.47352          1          1  ...   
3        제주국제공항(600번)  33.50577  126.49252          0         17  ...   
9                서귀포항  33.23977  126.56450          0          0  ...   
...               ...       ...        ...        ...        ...  ...   
415412  화순환승정류장(안덕농협)  33.24652  126.33507          0          1  ...   
415414        제주민속오일장  33.49376  126.47670          1          0  ...   
415418   한림환승정류장(한림리)  33.41437  126.26336          4          0  ...   
415419      제주시외버스터미널  33.49946  126.51479          4          0  ...   
415421   애월환승정류장(애월리)  33.46483  126.31870          1          0  ...   

        X9.10_takeoff  X10.11_takeoff  X11.12_takeoff  ride_total  \
0                   0               0               0          16   
1                   0               0               0          22   
2                   0               0               0           4   
3                   0               0               0          79   
9                   2               0               2           0   
...               ...             ...             ...         ...   
415412              0               0               0           1   
415414              0               0               0           1   
415418              0               0               0           4   
415419              0               0               0           4   
415421              0               0               0           1   

        takeoff_total  holiday  type_general_geton  type_other_geton  \
0                   0        0                15.0               1.0   
1                   0        0                22.0               0.0   
2                   0        0                 4.0               0.0   
3                   0        0                79.0               0.0   
9                   4        0                 0.0               0.0   
...               ...      ...                 ...               ...   
415412              0        1                 1.0               0.0   
415414              0        1                 1.0               0.0   
415418              0        1                 4.0               0.0   
415419              0        1                 4.0               0.0   
415421              0        1                 1.0               0.0   

        type_general_getoff  type_other_getoff  
0                       NaN                NaN  
1                       NaN                NaN  
2                       NaN                NaN  
3                       NaN                NaN  
9                       6.0                0.0  
...                     ...                ...  
415412                  NaN                NaN  
415414                  NaN                NaN  
415418                  NaN                NaN  
415419                  NaN                NaN  
415421                  NaN                NaN  

[195157

하차태그를 하지 않은 승객이 있어서  
일반 승객 하차인원 + 나머지 승객 하차인원 값이 총 하차인원과 다를 수 있다.